lib/ansible/modules/cloud/podman/podman_container.py

```diff
@@ -0,0 +1,1196 @@
+
+    def recreate(self):
+        self.delete()
+        self.run()
+
+    def restart(self):
+        self.stop()
+        self.run()
+
+
+class PodmanManager:
+
+    def __init__(self, module, results):
+
+        super(PodmanManager, self).__init__()
+
+        self.module = module
+        self.results = results
+        self.name = self.module.params.get('name')
+        self.executable = self.module.get_bin_path(module.params.get('executable'), required=True)
+        self.image = self.module.params.get('image')
+        if self.image:
+            if ensure_image_exists(self.module, self.image):
+                self.results['actions'].append('pulled image %s' % self.image)
+        self.state = self.module.params.get('state')
+        self.restart = self.module.params.get('force_restart')
+        self.recreate = self.module.params.get('recreate')
+        self.container = PodmanContainer(self.module, self.name, self.module.params)
+
+    def update_container_result(self, changed=True):
+        facts = self.container.get_info()
+        self.results.update({'changed': changed, 'container': facts,
+                             'ansible_facts': {'podman_container': facts}})
+        return self.results
+
+    def execute(self):
+        if self.state in ['present', 'started']:
+            if not self.container.exists and not self.image:
+                self.module.fail_json(msg='Cannot create container when image is not specified!')
+            if self.container.running and (self.container.different or self.recreate):
+                if not self.image:
+                    return self.fail(msg='Cannot create container when image is not specified!')
+                self.container.recreate()
+                self.results['actions'].append('recreated %s' % self.container.name)
+                return self.update_container_result()
+            elif self.container.running and not self.container.different:
+                if self.restart:
+                    self.container.restart()
+                    self.results['actions'].append('restarted %s' % self.container.name)
+                    return self.update_container_result()
+                return self.results
+            elif self.container.stopped and self.container.different:
+                self.container.recreate()
+                self.results['actions'].append('recreated %s' % self.container.name)
+                return self.update_container_result()
+            elif self.container.stopped and not self.container.different:
+                self.container.start()
+                self.results['actions'].append('started %s' % self.container.name)
+                return self.update_container_result()
+            elif not self.container.exists:
+                self.container.run()
+                self.results['actions'].append('started %s' % self.container.name)
+                return self.update_container_result()
+
+        if self.state == 'stopped':
```

Comment:
I'd probably separate all of the states into their own methods. And then in this method, you'd keep only an action router. It could look like this:
```python
        action_map = {
        	'present': self.make_started,
        	'started': self.make_started,
        	'stopped': self.make_stopped,
        	'absent': self.make_stopped,
        }

        try:
        	process_action = action_map[self.state]
        except KeyError:
        	return {
        		'changed': False, 'fail': True,
                'msg': 'Unexpected logic error happened, '
                'please contact maintainers ASAP!',
            }
        return process_action()
```

Reply:
great suggestion, will try to do that

Reply:
I check states in input params, so shouldn't be key error